### PR TITLE
snippet: Add many usefull snippets

### PR DESCRIPTION
**Pull request by Kerollmops**
These snippets comes from https://github.com/editor-rs/vscode-rust

### CODE REVIEW

**Comment by nrc** [comment]
Thanks for the PR! I would like to extend the snippets we have available, but I'm a bit cynical about adding every possible snippet en mass (and I'm not sure that taking these snippets as-is from vscode-rust without checking they are optimal). Are there any snippets you particularly miss having? Or that you feel would be high impact to have here?

**Comment by Kerollmops** [comment]
Ahah yes I'm missing `impl Trait for Xxx`, `fn`, `struct`, `struct ()`, `struct ;` and probably others.

I looked at the `impl Trait for Xxx` that can be dynamic, vscode doesn't seem to handle this for the moment https://github.com/Microsoft/vscode/issues/6920, https://github.com/rust-lang-nursery/rls-vscode/issues/100.

**Comment by Object905** [comment]
Snippets for attributes with prefix without `#![` can be really annoying, because when you want to write some attribute for example `#![feature(..)]` you start by writing `#![`, then `featu...` and snippet with prefix `feature` will show up and if you choose it you end up with something like `#![#![feature(..)]`. Maybe change prefix of such snippets to `#![keyword`. And it also would be useful to duplicate for `#[keyword` for attributes on a single item.

File: snippets/rust.json [hunk 1]
```diff
@@ -40,6 +453,117 @@
...
+            "    ${1:unimplemented!();}",
+            "}"
+        ],
+        "description": "fn main() { … }"
+    },
+    "match": {
+        "prefix": "match",
+        "body": [
+            "match ${1:expr} {",
+            "    ${2:Some(expr)} => ${3:expr},",
```
Comment (suhr) [inline snippets/rust.json]: Should be just `"    $2"`

File: snippets/rust.json [hunk 2]
```diff
@@ -40,6 +453,117 @@
...
+            "struct ${1:Name};"
+        ],
+        "description": "struct …;"
+    },
+    "struct": {
+        "prefix": "struct",
+        "body": [
+            "#[derive(Debug)]",
+            "struct ${1:Name} {",
+            "    ${2:field}: ${3:Type}",
```
Comment (suhr) [inline snippets/rust.json]: Should be just `"    $2"`

File: snippets/rust.json [hunk 3]
```diff
@@ -1,33 +1,446 @@
...
+        "prefix": "extern-crate",
+        "body": [
+            "extern crate ${1:name};"
+        ],
+        "description": "extern crate …;"
+    },
+    "extern-fn": {
+        "prefix": "extern-fn",
+        "body": [
+            "extern \"C\" fn ${1:name}(${2:arg}: ${3:Type}) -> ${4:RetType} {",
```
Comment (suhr) [inline snippets/rust.json]: Should be `"extern \"C\" fn ${1:name}($2) $3 {"`

File: snippets/rust.json [hunk 4]
```diff
@@ -1,33 +1,446 @@
...
+            "extern \"C\" {",
+            "    ${2:// add code here}",
+            "}"
+        ],
+        "description": "extern \"C\" { … }"
+    },
+    "fn": {
+        "prefix": "fn",
+        "body": [
+            "fn ${1:name}(${2:arg}: ${3:Type}) -> ${4:RetType} {",
```
Comment (suhr) [inline snippets/rust.json]: Should be `"fn ${1:name}($2) $3 {`

File: snippets/rust.json [hunk 5]
```diff
@@ -1,33 +1,446 @@
...
+        "description": "stringify!(…)"
+    },
+    "thread_local": {
+        "prefix": "thread_local",
+        "body": [
+            "thread_local!(static ${1:STATIC}: ${2:Type} = ${4:init});"
+        ],
+        "description": "thread_local!(static …: … = …);"
+    },
+    "try": {
```
Comment (suhr) [inline snippets/rust.json]: Do we still need `try!`?

File: snippets/rust.json [hunk 6]
```diff
@@ -1,33 +1,446 @@
...
+            "}"
+        ],
+        "description": "else { … }"
+    },
+    "enum": {
+        "prefix": "enum",
+        "body": [
+            "#[derive(Debug)]",
+            "enum ${1:Name} {",
+            "    ${2:Variant1},",
```
Comment (suhr) [inline snippets/rust.json]: Should be just `"    $2"`

File: snippets/rust.json (hunk 5 above)
Comment (Kerollmops) [reply snippets/rust.json]: nope

File: snippets/rust.json (hunk 6 above)
Comment (Kerollmops) [reply snippets/rust.json]: IFAIK the "Variant1" text is just the default text placed, no ?

File: snippets/rust.json (hunk 3 above)
Comment (Kerollmops) [reply snippets/rust.json]: even the `RetType` doesn't need a default name

Comment (Kerollmops) [reply snippets/rust.json]: Ho yes sorry, I didn't understand your change proposition

File: snippets/rust.json (hunk 6 above)
Comment (Kerollmops) [reply snippets/rust.json]: yes but in some cases it's just a hint for the user

Comment (Kerollmops) [reply snippets/rust.json]: and you have just to write hover it

**Comment by Kerollmops** [comment]
Up on this PR ?

**Comment by nrc** [comment]
Sorry, perhaps I was unclear with my earlier comment. This PR seems to add every snippet from vscode-rust, I don't want to do that. I am not convinced there is value in it (and there are downsides in having so many snippets - it can be confusing for users and contributors, and it can make code completion worse).

So, I would rather add a few snippets at a time ensuring that each snippet is well-motivated and optimally specified.

**Pull request closed without merging**